### PR TITLE
LibCore: FilePermissionsMask

### PR DESCRIPTION
**Pull request by xavier**
This change introduces a Unix file permission parser class as discussed in https://github.com/SerenityOS/serenity/pull/11319.

I have rewritten the symbolic permission parsing algorithm to be FSM-based. It should now be more straightforward as it reduces the number of hoops and doesn't involve recursive calls anymore.

I have updated `mkdir` and `chmod` to use the new library, hence reducing the complexity of both utilities.

### CODE REVIEW

**Review by linusg — CHANGES_REQUESTED** [review]
Please:

- Use east const
- Wrap commit messages at 72 characters, not 50

...as per the [contribution](https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy) and [coding style](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#const-placement) guidelines.

File: Userland/Libraries/LibCore/FilePermissionsMask.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    User = 4
+};
+
+enum Operation {
+    Add,
+    Remove,
+    Assign,
+};
+
+ErrorOr<FilePermissionsMask> FilePermissionsMask::parse(const StringView& string)
```
Comment (linusg) [inline Userland/Libraries/LibCore/FilePermissionsMask.cpp:32]: Don't pass StringView by const reference, pass it by value (see https://github.com/SerenityOS/serenity/commit/8b1108e4858f797c9216dc8ae4a3918ad50c73b4)

File: Userland/Libraries/LibCore/FilePermissionsMask.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#include <AK/Error.h>
+#include <AK/String.h>
+#include <sys/stat.h>
+
+namespace Core {
+
+class FilePermissionsMask {
+private:
+    mode_t m_clear_mask; // the bits that will be cleared
+    mode_t m_write_mask; // the bits that will be set
```
Comment (linusg) [inline Userland/Libraries/LibCore/FilePermissionsMask.h:18]: We use public, protected, private as the usual order of class members and functions

**Comment by roytam1** [comment]
it seems break normal usage by scripts, getting this when running perl's Configure script:
```
40.233 chmod(127:127): Exiting with runtime error: invalid access class: expected 'u', 'g', 'o' or 'a'
```

**Comment by xavier** [comment]
@roytam1 Do you by any chance have an example of the symbolic permission which causes this problem? Or a way to reproduce it?

**Comment by bgianfo** [comment]
> it seems break normal usage by scripts, getting this when running perl's Configure script:
> 
> ```
> 40.233 chmod(127:127): Exiting with runtime error: invalid access class: expected 'u', 'g', 'o' or 'a'
> ```

Please file a bug with directions on how to reproduce.

**Comment by roytam1** [comment]
> @roytam1 Do you by any chance have an example of the symbolic permission which causes this problem? Or a way to reproduce it?

a simple `chmod +r README.md` will do, and mode bits are intact.
IIRC without access class implies `a`.

**Comment by xavier** [comment]
Gotcha, I'll work on a fix.

**Pull request merged**